### PR TITLE
Use Rust HTTPS Everywhere library

### PR DESCRIPTION
**Pull request by antonok-edm**
Resolves https://github.com/brave/brave-browser/issues/5280.

Blocked on acceptance of https://github.com/brave/https-everywhere-builder/pull/40 to ship the new format accepted by the rust library.

I've generated a private key, public key, and component ID for the new v7 component according to the instructions on https://github.com/brave/brave-core-crx-packager#development, however this is not currently uploaded anywhere because I don't have permissions to do so. If it's preferred for someone else to generate/upload these keys I am happy to update this PR.

Users who are upgrading from previous versions of Brave will automatically uninstall the v6 component and upgrade to the new v7 component; users who are installing for the first time or have already migrated will just load the v7 component normally.

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

The following steps should be tested:
 - Without any existing profile directory
 - With an existing profile from a previous version of Brave
 - With an existing profile from this version of Brave (i.e. one of the above after a browser restart, just be sure to re-enable HTTPS Everywhere in the Shields panel)

1. On Desktop, navigate to http://https-everywhere.badssl.com.
2. Ensure the page appears with a checkmark at the bottom that says "HTTPS Everywhere is working".
3. Ensure that the Shields icon shows 1 on the counter. Opening the Shields panel in advanced mode should show 1 connection upgraded to HTTPS. Clicking the dropdown on the 1 should show the single entry, `https-everywhere.badssl.com`.
4. Disabling the "Connections upgraded to HTTPS" toggle in Shields should result in the page refreshing. After the reload, the checkmark should be replaced by a red X, labeled "HTTPS Everywhere is not working".
5. Close the browser and verify that the profile directory does _not_ contain a directory named `oofiananboodjbbmdelgdommihjbkfag`.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

**Review by fmarier — APPROVED** [review]
The change in the license script looks good.

File: components/brave_shields/browser/https_everywhere_service.cc [hunk 1]
```diff
@@ -12,83 +12,70 @@
...
-#include "base/values.h"
-#include "third_party/leveldatabase/src/include/leveldb/db.h"
-#include "third_party/re2/src/re2/re2.h"
+#include "brave/browser/component_updater/brave_component_installer.h"
+#include "brave/common/pref_names.h"
+#include "brave/vendor/https-everywhere-lib-cpp/src/wrapper.hpp"
+#include "chrome/browser/browser_process_impl.h"
+#include "chrome/browser/first_run/first_run.h"
+#include "components/component_updater/component_updater_service.h"
+#include "components/prefs/pref_service.h"
```
Comment (antonok-edm) [inline components/brave_shields/browser/https_everywhere_service.cc:26]: I'll need to add `//chrome/browser` and `//brave/browser` to `components/brave_shields/browser/BUILD.gn` for these, but both cause a dependency cycle.
@bridiver any ideas?

File: components/brave_shields/browser/https_everywhere_service.h [hunk 2]
```diff
@@ -16,10 +16,8 @@
 #include "base/synchronization/lock.h"
 #include "brave/components/brave_shields/browser/base_brave_shields_service.h"
 #include "brave/components/brave_shields/browser/https_everywhere_recently_used_cache.h"
-
-namespace leveldb {
-class DB;
-}
+#include "brave/vendor/https-everywhere-lib-cpp/src/wrapper.hpp"
```
Comment (bridiver) [inline components/brave_shields/browser/https_everywhere_service.h:19]: the `.hpp` convention is deprecated. Even if this stays as an external repo we should follow the conventions (paths, etc...) from speedreader

File: components/brave_shields/browser/https_everywhere_service.cc (hunk 1 above)
Comment (bridiver) [reply components/brave_shields/browser/https_everywhere_service.cc:26]: this is not allowed. You should not be using `g_browser_process` and other things from `//chrome/browser`, `//brave/browser` or `//brave/common`. 
As an example of how to refactor these deps https://github.com/brave/brave-core/pull/6520/files#diff-468a5d483ee258ec61c8b136e8123e16R98 you should pass the local state into the HTTPSEverywhereService constructor

File: components/brave_shields/browser/https_everywhere_service.cc [hunk 3]
```diff
@@ -100,15 +87,29 @@ HTTPSEverywhereService::g_https_everywhere_component_base64_public_key_(
...
-  GetTaskRunner()->DeleteSoon(FROM_HERE, level_db_);
 }
 
 bool HTTPSEverywhereService::Init() {
+  PrefService* local_state = g_browser_process->local_state();
+  if (!local_state) {
+    return false;
+  }
+
+  bool hasMigratedHTTPSEverywhere6 =
```
Comment (bridiver) [inline components/brave_shields/browser/https_everywhere_service.cc:103]: there is actually a specific place where this is supposed to happen - you should do a chromium_src override for `MigrateObsoleteLocalStatePrefs` from chrome/browser/prefs/browser_prefs.cc

File: components/brave_shields/browser/https_everywhere_service.cc [hunk 4]
```diff
@@ -12,83 +12,70 @@
...
 }
-std::string leveldbGet(leveldb::DB* db, const std::string &key) {
-  if (!db) {
-    return "";
-  }
 
-  std::string value;
-  leveldb::Status s = db->Get(leveldb::ReadOptions(), key, &value);
-  return s.ok() ? value : "";
+void MigrateHTTPSEverywhere6Prefs() {
```
Comment (bridiver) [inline components/brave_shields/browser/https_everywhere_service.cc:54]: the service shouldn't be responsible for the component, this code should go in `BraveComponentLoader`

File: components/brave_shields/browser/https_everywhere_service.cc [hunk 5]
```diff
@@ -117,30 +118,20 @@ bool HTTPSEverywhereService::Init() {
...
-                        &level_db_);
-  if (!status.ok() || !level_db_) {
-    LOG(ERROR) << "Level db open error "
-               << unzipped_level_db_path.value().c_str()
-               << ", error: " << status.ToString();
-    CloseDatabase();
-    return;
-  }
+  std::string rules;
+  base::ReadFileToString(json_ruleset_path, &rules);
```
Comment (bridiver) [inline components/brave_shields/browser/https_everywhere_service.cc:133]: how big is this list?

File: components/brave_shields/browser/https_everywhere_service.cc [hunk 6]
```diff
@@ -117,30 +118,20 @@ bool HTTPSEverywhereService::Init() {
...
+  rust_client_.LoadRules(rules);
```
Comment (bridiver) [inline components/brave_shields/browser/https_everywhere_service.cc:134]: does the rust lib account for the fact that `LoadRules` happens on a different thread than `RewriteUrl`? I'm a bit concerned in general about managing threads in rust vs c++ and I think we should talk about this in DM so I can understand what's going on when the rules are loaded.

File: browser/brave_local_state_prefs.cc [hunk 7]
```diff
@@ -43,6 +44,7 @@ namespace brave {
 
 void RegisterLocalStatePrefs(PrefRegistrySimple* registry) {
   brave_shields::RegisterPrefsForAdBlockService(registry);
+  brave_shields::RegisterPrefsForHTTPSEverywhereService(registry);
```
Comment (bridiver) [inline browser/brave_local_state_prefs.cc:47]: this should be `HTTPSEverywhereService::RegisterLocalStatePrefs`

File: common/pref_names.cc [hunk 8]
```diff
@@ -41,6 +41,7 @@ const char kReferralHeaders[] = "brave.referral.headers";
 const char kReferralAndroidFirstRunTimestamp[] =
     "brave.referral_android_first_run_timestamp";
 const char kHTTPSEVerywhereControlType[] = "brave.https_everywhere_default";
+const char kMigratedHTTPSEverywhere6[] = "brave.https_everywhere.migrated6";
```
Comment (bridiver) [inline common/pref_names.cc:44]: these should go in `components/brave_shields/browser/pref_names`

File: components/brave_shields/browser/BUILD.gn [hunk 9]
```diff
@@ -51,12 +51,15 @@ source_set("browser") {
 
   deps = [
     "//base",
+    "//brave/common",
```
Comment (bridiver) [inline components/brave_shields/browser/BUILD.gn:54]: this is not allowed

File: components/brave_shields/browser/https_everywhere_service.cc [hunk 10]
```diff
@@ -100,15 +87,29 @@ HTTPSEverywhereService::g_https_everywhere_component_base64_public_key_(
 HTTPSEverywhereService::HTTPSEverywhereService(
     BraveComponent::Delegate* delegate)
     : BaseBraveShieldsService(delegate),
-      level_db_(nullptr) {
+      rust_client_(httpse::HttpsEverywhereClient()) {
```
Comment (bridiver) [inline components/brave_shields/browser/https_everywhere_service.cc:90]: you don't have to initialize non-pointer class members, but I think this should probably be a unique_ptr

**Comment by diracdeltas** [comment]
any updates on this?

**Comment by antonok-edm** [comment]
@diracdeltas there were some concerns from code review about additional memory usage - the EFF library holds rules (for reference, it's about 5MB of JSON) in-memory, whereas the leveldb implementation we currently use is kept purely on disk. I proposed a modular storage backend, but @bridiver wants to avoid file IO from our rust libraries. There isn't really a clear path on how to proceed at the moment.

**Comment by diracdeltas** [comment]
got it. i was mainly wondering if this work would make it easier to enable continual ruleset updates (https://www.eff.org/deeplinks/2018/04/https-everywhere-introduces-new-feature-continual-ruleset-updates).

unless something has changed since the last time i looked at the repo, we still update manually using https-everywhere-builder and are several months behind EFF's ruleset library.

**Comment by bbondy** [comment]
Just noting we run adblock rule updates on Jenkins every day, so there are daily updates. We could do the same for existing packaging for HTTPSE.

**Comment by diracdeltas** [comment]
@bbondy that would actually be awesome. https://github.com/brave/https-everywhere-builder should work for this since `npm run build` in it fetches the latest rulesets from https://www.https-rulesets.org/v1/ and builds them

**Comment by bbondy** [comment]
@diracdeltas I think the best way to do that is to add an issue similar to this https://github.com/brave/devops/issues/3419 in the `devops` repo.

**Comment by bbondy** [comment]
No activity since 2020, closing.

**Pull request closed without merging**